types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,423 @@
+	wl_list_remove(&constraint->focus_change.link);
+	if (constraint->pointer_motion.link.next) {
+		wlr_log(L_DEBUG, "pointer_motion listeners uninstalled");
+		wl_list_remove(&constraint->pointer_motion.link);
+		wl_list_remove(&constraint->touch_motion.link);
+	}
+	free(constraint);
+}
+
+static void pointer_constraint_destroy_resource(struct wl_resource *resource) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	pointer_constraint_destroy(constraint);
+}
+
+static void pointer_constraint_set_region(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *region) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.region = region;
+}
+
+// only valid for locked_pointer
+static void pointer_constraint_set_cursor_position_hint(struct wl_client *client,
+		struct wl_resource *resource, wl_fixed_t x, wl_fixed_t y) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.cursor_hint.x = wl_fixed_to_double(x);
+	constraint->pending.cursor_hint.y = wl_fixed_to_double(y);
+	constraint->pending.cursor_hint.valid = true;
+}
+
+static void handle_focus_change(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, focus_change);
+
+	struct wlr_seat *seat = data;
+	if (seat != constraint->seat) {
+		return;
+	}
+
+	send_unconstrained(constraint);
+	if (constraint->seat->pointer_state.focused_surface == constraint->surface) {
+		wlr_log(L_DEBUG, "found constraint %p for %p", constraint, seat);
+		pointer_constraint_check_no_region_and_activate(constraint);
+	}
+}
+
+// checks whether we are within the region
+static void handle_motion(struct wlr_pointer_constraint_v1 *constraint, void *data) {
+	/*
+	 * If our constraint is not accepted, or does not have a specific
+	 * region, or is already active, we return.
+	 *
+	 * Handling of NULL regions is done in pointer_constraint_check_no_region_and_activate.
+	 */
+	if (constraint->seat->pointer_state.focused_surface != constraint->surface ||
+			!constraint->accepted ||
+			!constraint->current.region ||
+			constraint->active) {
+		wlr_log(L_DEBUG, "pointer_motion listeners uninstalled");
+		wl_list_remove(&constraint->pointer_motion.link);
+		wl_list_remove(&constraint->touch_motion.link);
+		return;
+	}
+
+	struct wlr_seat_pointer_motion_event *event = data;
+
+	pixman_box32_t pbox;
+	if (pixman_region32_contains_point(
+			wlr_region_from_resource(constraint->current.region),
+			event->sx, event->sy, &pbox)) {
+		struct wlr_box box = WLR_BOX_EMPTY;
+		if (constraint->type == WLR_POINTER_CONSTRAINT_V1_CONFINED) {
+			wlr_box_from_pixman_box32(pbox, &box);
+		}
+
+		struct wlr_pointer_constraint_v1_activation_event activation = {
+			.pointer_constraint = constraint,
+			.box = box,
```

Comment:
It looks like you're only sending and using only one box in the region, but there can be many boxes in the region.

Reply:
I'm only sending the box which the cursor is currently within, but now that I think about it, this is probably wrong. If you have two adjacent boxes within the region, you wouldn't be able to pass through from one box to the other.